test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala

```diff
@@ -0,0 +1,163 @@
+import org.junit.runners.JUnit4
+
+import scala.reflect.internal.util._
+
+
+@RunWith(classOf[JUnit4])
+class ConsoleReporterTest {
+  val source = "Test_ConsoleReporter"
+  val batchFile = new BatchSourceFile(source, "For testing".toList)
+  val posWithSource = new OffsetPosition(batchFile, 4)
+  val content = posWithSource.lineContent
+  val writerOut = new ByteArrayOutputStream()
+  val echoWriterOut = new ByteArrayOutputStream()
+  val reporter = createConsoleReporter("s", writerOut, echoWriterOut)
+
+  /** Change maxerrs and maxwarns */
+  reporter.settings.maxerrs.value = 1
+  reporter.settings.maxwarns.value = 1
+
+
+  def createConsoleReporter(inputForReader: String, writerOut: ByteArrayOutputStream, echoWriterOut: ByteArrayOutputStream = null): ConsoleReporter = {
+      val writer = new PrintWriter(writerOut)
+      val reader = new BufferedReader(new StringReader(inputForReader))
+      
+      /** Create reporter with the same writer and echoWriter if None */
+      Option(echoWriterOut) match {
+        case None => new ConsoleReporter(new Settings(), reader, writer)
+        case Some(_) => new ConsoleReporter(new Settings(), reader, writer, new PrintWriter(echoWriterOut))
+      }
+  }
+
+
+  def testHelper(pos: Position = NoPosition, msg: String, severity: String = "")(test: Position => Unit) = {
+    test(pos)
+    if (msg.isEmpty && severity.isEmpty) assertTrue(writerOut.toString.isEmpty)
+    else {
+      if (!pos.isDefined) assertEquals(severity + msg, writerOut.toString.lines.next)
+      else {
+        val it = writerOut.toString.lines
+        assertEquals(source + ":1: " + severity + msg, it.next)
+        assertEquals(content, it.next)
+        assertEquals("    ^", it.next)
+      }
+    }
+    writerOut.reset
+  }
+
+  
+  @Test
+  def printMessageTest(): Unit = {
+    testHelper(msg = "Hello World!")(_ => reporter.printMessage("Hello World!"))
+    testHelper(msg = "Testing with NoPosition")(_ => reporter.printMessage(NoPosition, "Testing with NoPosition"))
+    testHelper(posWithSource, "Testing with Defined Position")(_ => reporter.printMessage(posWithSource, "Testing with Defined Position"))
+  }
+
+
+  @Test
+  def echoTest(): Unit = {
+    reporter.echo("Hello World!")
+    assertEquals("Hello World!", echoWriterOut.toString.lines.next)
+
+    /** Check with constructor which has the same writer and echoWriter */
+    val reporter2 = createConsoleReporter("r", writerOut)
+    testHelper(msg = "Hello World!")(_ => reporter2.echo("Hello World!"))
+  }
+
+
+  @Test
+  def printTest(): Unit = {
+    testHelper(msg = "test")(_ => reporter.print(NoPosition, "test", reporter.INFO))
+    testHelper(msg = "test", severity = "warning: ")(_ => reporter.print(NoPosition, "test", reporter.WARNING))
+    testHelper(msg = "test", severity = "error: ")(_ => reporter.print(NoPosition, "test", reporter.ERROR))
+    testHelper(posWithSource, msg = "test")(_ => reporter.print(posWithSource, "test", reporter.INFO))
```

Comment:
Since your helper takes `Pos => Unit`, this would be
```
testHelper(posWithSource, msg = "test")(reporter.print(_, "test", reporter.INFO))
```
If you find that finicky, then helper could just take `(body: => Unit)`, and maybe that's easier to read, too:
```
testHelper(posWithSource, msg = "test")(reporter.print(posWithSource, "test", reporter.INFO))
```